### PR TITLE
Add SubjectMetadataController

### PR DESCRIPTION
**Pull request by rekmarks**
The extension permission metadata functionality is currently housed within its local permissions controller. This PR extracts this functionality into a new "subject" (previously known as "domain") metadata controller, per [the extension implementation](https://github.com/MetaMask/metamask-extension/blob/0d63aa04021142a18b7e1aefce9d060ece364340/app/scripts/controllers/permissions/index.js/#L521-L563).

### CODE REVIEW

**Comment by rekmarks** [comment]
We should merge this PR into `main` once the new permissions controller has been merged. Also TBD whether we're going to implement `metamask_sendDomainMetadata` as part of this PR or do something else with that.

File: packages/controllers/src/subject-metadata/SubjectMetadataController.ts [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+   * method is static because we want to call it in the constructor, before the
+   * controller's state is initialized.
+   *
+   * @param draftState - The `immer` draft state.
+   * @param hasPermissions - A function that returns a boolean indicating
+   * whether a particular subject (identified by its origin) has any
+   * permissions.
+   */
+  private static trimMetadataState(
+    draftState: SubjectMetadataControllerState,
```
Comment (Gudahtt) [inline packages/controllers/src/subject-metadata/SubjectMetadataController.ts:250]: The parameter is called `draftState` and it's documented as being an Immer draft, but the type contradicts this. It's typed as the state, not as a draft. It's also used this way in the constructor, where the state is passed in here directly (not as a draft).

Comment (Gudahtt) [reply packages/controllers/src/subject-metadata/SubjectMetadataController.ts:250]: I think it would be simpler for this to return the trimmed state, rather than mutate the input. Then it doesn't matter whether what's passed in is a draft or not. We could then use `castImmutable` when passing in the draft, so that this doesn't need to accept a Draft type, letting us use this more easily in the constructor as well.

File: packages/controllers/src/subject-metadata/SubjectMetadataController.ts [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+type SubjectOrigin = string;
+
+export type SubjectMetadata = PermissionSubjectMetadata & {
+  [key: string]: Json;
+  name: string;
+  // TODO:TS4.4 make optional
+  extensionId: string | null;
+  host: string | null;
+  iconUrl: string | null;
```
Comment (Gudahtt) [inline packages/controllers/src/subject-metadata/SubjectMetadataController.ts:24]: This was called `icon` previously. Was this change intentional?

I suppose this is a better name, but this change will require a migration.

File: packages/controllers/src/subject-metadata/SubjectMetadataController.ts [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+      );
+    }
+
+    const hasPermissions = (origin: string) => {
+      return messenger.call('PermissionController:hasPermissions', origin);
+    };
+
+    if (state.subjectMetadata) {
+      SubjectMetadataController.trimMetadataState(
+        state as SubjectMetadataControllerState,
```
Comment (Gudahtt) [inline packages/controllers/src/subject-metadata/SubjectMetadataController.ts:125]: This cast doesn't seem correct because this state is `Partial<SubjectMetadataControllerState>`, not `SubjectMetadataControllerState`.

It's true that `subjectMetadata` is definitely truthy, and that's the only property that we're using, but if other properties are defined in the future this could lead to errors.

As an alternative, can we update `trimMetadataState` to only accept the `subjectMetadata` object? The types would be easier that way.

File: packages/controllers/src/subject-metadata/SubjectMetadataController.ts [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+   * controller.
+   */
+  private registerMessageHandlers(): void {
+    this.messagingSystem.registerActionHandler(
+      `${controllerName}:clearState`,
+      () => this.clearState(),
+    );
+
+    this.messagingSystem.registerActionHandler(
+      `${controllerName}:trimMetadataState`,
```
Comment (Gudahtt) [inline packages/controllers/src/subject-metadata/SubjectMetadataController.ts:154]: I can't think of a reason another controller would have for wanting to trim the metadata state. What did you have in mind for this action?

File: packages/controllers/src/subject-metadata/SubjectMetadataController.ts [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+  GenericPermissionController,
+  PermissionSubjectMetadata,
+  HasPermissions,
+} from '../permissions';
+
+const controllerName = 'SubjectMetadataController';
+
+type SubjectOrigin = string;
+
+export type SubjectMetadata = PermissionSubjectMetadata & {
```
Comment (Gudahtt) [inline packages/controllers/src/subject-metadata/SubjectMetadataController.ts:18]: The old subject metadata included a `lastUpdated` property. Was that removed intentionally? I'm not sure what that was intended for to be honest. It appears to be unused at the moment.

File: packages/controllers/src/subject-metadata/SubjectMetadataController.ts [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+
+    let originToForget: string | null = null;
+    // We only delete the oldest encountered subject from the cache, again to
+    // ensure that the user's experience isn't degraded by missing icons etc.
+    if (this.subjectsEncounteredSinceStartup.size >= this.subjectCacheLimit) {
+      const cachedOrigin = this.subjectsEncounteredSinceStartup
+        .values()
+        .next().value;
+
+      this.subjectsEncounteredSinceStartup.delete(cachedOrigin);
```
Comment (Gudahtt) [inline packages/controllers/src/subject-metadata/SubjectMetadataController.ts:197]: So wait.... we're removing the origin from this list, even if we want to preserve the metadata? Why would we let these two data sets get out of sync like that?

Comment (Gudahtt) [reply packages/controllers/src/subject-metadata/SubjectMetadataController.ts:197]: I see that the existing subject metadata controller works basically the same way, and it seems like a bug there as well.

File: packages/controllers/src/subject-metadata/SubjectMetadataController.ts [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+
+    if (
+      !newMetadata.extensionId &&
+      !newMetadata.host &&
+      /[a-z]+:\/\/\w.+/iu.test(origin)
+    ) {
+      try {
+        newMetadata.host = new URL(origin).host;
+      } catch (_) {
+        // Do nothing. Some "origins" are not valid URLs.
```
Comment (Gudahtt) [inline packages/controllers/src/subject-metadata/SubjectMetadataController.ts:213]: What does this refer to? The only invalid origin I would expect here is `metamask`. Ignoring this error in general seems like a bad idea.

File: packages/controllers/src/subject-metadata/SubjectMetadataController.ts [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+      if (!this.subjectHasPermissions(cachedOrigin)) {
+        originToForget = cachedOrigin;
+      }
+    }
+
+    this.subjectsEncounteredSinceStartup.add(origin);
+
+    if (
+      !newMetadata.extensionId &&
+      !newMetadata.host &&
```
Comment (Gudahtt) [inline packages/controllers/src/subject-metadata/SubjectMetadataController.ts:207]: Perhaps we should always set the host ourselves, rather than trusting the caller to set it properly? If we have the origin and are guaranteeing the accuracy of that ourselves, it seems better to always use that. Otherwise the caller could lie about the host.

File: packages/controllers/src/subject-metadata/SubjectMetadataController.ts (hunk 4 above)
Comment (Gudahtt) [reply packages/controllers/src/subject-metadata/SubjectMetadataController.ts:154]: I can't think of an obvious purpose for `clearState` either.

File: packages/controllers/src/subject-metadata/SubjectMetadataController.ts [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+ * or less, a cache.
+ */
+export class SubjectMetadataController extends BaseController<
+  typeof controllerName,
+  SubjectMetadataControllerState,
+  SubjectMetadataControllerMessenger
+> {
+  private _subjectCacheLimit: number;
+
+  get subjectCacheLimit(): number {
```
Comment (Gudahtt) [inline packages/controllers/src/subject-metadata/SubjectMetadataController.ts:100]: What is this for?

File: packages/controllers/src/subject-metadata/SubjectMetadataController.ts [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+      () => this.trimMetadataState(),
+    );
+  }
+
+  /**
+   * Clears the state of this controller.
+   */
+  clearState(): void {
+    this.update((_draftState) => {
+      return { ...defaultState };
```
Comment (Gudahtt) [inline packages/controllers/src/subject-metadata/SubjectMetadataController.ts:164]: Letting this get out of sync with `subjectsEncounteredSinceStartup` seems bad. This could lead to a situation where the actual state is getting trimmed to zero because a `clear` left domains in `subjectsEncounteredSinceStartup` equaling the limit.

File: packages/controllers/src/subject-metadata/SubjectMetadataController.ts (hunk 5 above)
Comment (rekmarks) [reply packages/controllers/src/subject-metadata/SubjectMetadataController.ts:18]: I had forgotten about this property, but I think we can get rid of it.

IIRC it was intended for the rosy future when we can fetch metadata on demand, and use `lastUpdated` to determine when to do so. However, since fetching metadata is cheap, we can probably just get rid of it?

File: packages/controllers/src/subject-metadata/SubjectMetadataController.ts (hunk 7 above)
Comment (rekmarks) [reply packages/controllers/src/subject-metadata/SubjectMetadataController.ts:213]: I intend for snap "origins" to get passed in here, e.g. `npm:@metamask/example-snap`. I suppose we could require that added subjects indicate whether they should have a valid origin or not, would that resolve the concern?

File: packages/controllers/src/subject-metadata/SubjectMetadataController.ts (hunk 9 above)
Comment (rekmarks) [reply packages/controllers/src/subject-metadata/SubjectMetadataController.ts]: Hrm, it's not necessary. Removed in: f7bd1cc

File: packages/controllers/src/subject-metadata/SubjectMetadataController.ts (hunk 7 above)
Comment (Gudahtt) [reply packages/controllers/src/subject-metadata/SubjectMetadataController.ts:209]: On second thought, maybe we don't need `host` at all? I was just trying to look up what we use it for, and I can barely find any uses. The one usage I can find is in `getPermissionsMetadataHostCounts`, which is only used to decide whether to strip `https` or not. I don't think we need it. We should always strip `HTTPS` if the origin is just `https://[hostname]`, and never strip it if the origin includes extra stuff like a port, a username/password, etc.

For snaps, we haven't discussed doing any equivalent shortening, so the `host` isn't relevant there either.

File: packages/controllers/src/subject-metadata/SubjectMetadataController.ts (hunk 2 above)
Comment (rekmarks) [reply packages/controllers/src/subject-metadata/SubjectMetadataController.ts:24]: Yeah, I'm planning on it!

File: packages/controllers/src/subject-metadata/SubjectMetadataController.ts (hunk 4 above)
Comment (rekmarks) [reply packages/controllers/src/subject-metadata/SubjectMetadataController.ts]: Yeah... Deleted both in 7a5aa3d

File: packages/controllers/src/subject-metadata/SubjectMetadataController.ts (hunk 7 above)
Comment (rekmarks) [reply packages/controllers/src/subject-metadata/SubjectMetadataController.ts:179]: Oh sweet I'm gonna nuke it.

Comment (rekmarks) [reply packages/controllers/src/subject-metadata/SubjectMetadataController.ts]: I nuked it: 03e4acc

File: packages/controllers/src/subject-metadata/SubjectMetadataController.ts (hunk 1 above)
Comment (rekmarks) [reply packages/controllers/src/subject-metadata/SubjectMetadataController.ts]: Yeah, agreed. I still had to do an `any` cast due to the "infinite type" bug and therefore didn't use `castImmutable`, but I think this is better: ec6a216

File: packages/controllers/src/subject-metadata/SubjectMetadataController.ts (hunk 3 above)
Comment (rekmarks) [reply packages/controllers/src/subject-metadata/SubjectMetadataController.ts]: Fixed in: 62a6af4

File: packages/controllers/src/subject-metadata/SubjectMetadataController.ts (hunk 10 above)
Comment (rekmarks) [reply packages/controllers/src/subject-metadata/SubjectMetadataController.ts:125]: Yeah, good point! fb306e8

File: packages/controllers/src/subject-metadata/SubjectMetadataController.ts (hunk 6 above)
Comment (rekmarks) [reply packages/controllers/src/subject-metadata/SubjectMetadataController.ts:157]: No, this is intentional. If a subject encountered since startup has permissions, the current implementation essentially says "screw it, if permissions are ever removed for this subject, we'll let the next subsequent call to `trimMetadataState` remove it". That'll likely occur at the next reboot of the extension, and that should be good enough for our purposes.

Comment (Gudahtt) [reply packages/controllers/src/subject-metadata/SubjectMetadataController.ts:157]: But why keep it around indefinitely if we don't need to? I don't see the advantage to doing this.

If we keep these two collections in sync, and only remove from the set when we remove the actual metadata, we can be confident that the trimming will work correctly and remove any excess metadata without associated permissions. As things stand now it could grow unbounded until the next reset.

File: packages/controllers/src/subject-metadata/SubjectMetadataController.test.ts [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+      });
+
+      controller.addSubjectMetadata(getSubjectMetadata('foo.com', 'foo'));
+
+      expect(controller.state).toStrictEqual({
+        subjectMetadata: { 'foo.com': getSubjectMetadata('foo.com', 'foo') },
+      });
+
+      expect(
+        (controller as any).subjectsEncounteredSinceStartup.size,
```
Comment (Gudahtt) [inline packages/controllers/src/subject-metadata/SubjectMetadataController.test.ts:118]: This test is looking at internals instead of asserting what it's meant to prove, which is that the trimming still works correctly.

Instead of doing this, you could test that after clearing state, it's capable of holding X new sites again, where X is the limit. That would confirm this aspect of clearing state works correctly, and it would be less coupled with the implementation.

Comment (rekmarks) [reply packages/controllers/src/subject-metadata/SubjectMetadataController.test.ts:118]: Ahhh, I had to try.

Comment (rekmarks) [reply packages/controllers/src/subject-metadata/SubjectMetadataController.test.ts]: Fixed here: 66d6eb5

File: packages/controllers/src/subject-metadata/SubjectMetadataController.ts (hunk 6 above)
Comment (rekmarks) [reply packages/controllers/src/subject-metadata/SubjectMetadataController.ts:157]: > As things stand now it could grow unbounded until the next reset.

Yes, but since it's been working like this for ~two years, we can be confident that it doesn't grow out of control in practice. The next reset will be no later than the next update, and our update frequency is if anything going to increase as opposed to decrease. As for users that don't update or restart their browsers, I think that's out of scope.

If if we were to implement your suggestion, unbounded growth is still theoretically possible since we never delete metadata for subjects with permissions under any circumstance.

> But why keep it around indefinitely if we don't need to? I don't see the advantage to doing this.

There are advantages, though! In your suggested implementation, we would have to iterate over the entire since-startup set if it's larger than the cache limit, check the permissions of each subject, and delete the first one we find without permissions. The unlucky case here is that the user adds permissions to every subject they encounter. In that case, there will eventually be a computational cost to performing this iteration every time the user visits a new website.

In addition, changing how this work can result in a degraded user experience. Consider again the case where the user adds permissions to every subject under your suggested implementation. Let's say, after hitting the cache limit, they visit a new website. By the time we receive a permissions request from that website, its metadata will have been deleted. The only way to add metadata for it and any subsequent new websites would be to accept the request (without metadata) and refresh the page.

This brings me to the final advantage, which is that improving the existing implementation is more complicated than it seems, and probably not worth our time given that the existing implementation demonstrably works in practice. I think the real solution to this problem is to be able to request metadata on demand, which requires the series of changes to `json-rpc-engine` we've discussed in the past.

Comment (Gudahtt) [reply packages/controllers/src/subject-metadata/SubjectMetadataController.ts:157]: >Consider again the case where the user adds permissions to every subject under your suggested implementation. Let's say, after hitting the cache limit, they visit a new website. By the time we receive a permissions request from that website, its metadata will have been deleted. The only way to add metadata for it and any subsequent new websites would be to accept the request (without metadata) and refresh the page.

Ok, fair enough, that makes sense. I still don't like the idea of letting it grow unbounded, and I do not accept the last two years as evidence that this isn't a problem (that does not follow). But you are right that my suggestion is bad and this is more complicated than I thought.

Could we at least rename the variable though? I expected `subjectsEncounteredSinceStartup` to be the subjects encountered since startup, and was very surprised to find that it was not that at all. `maybe something like `subjectsWithoutPermissionsEcounteredSinceStartup`? It's absurdly long but it's clear at least. Or `subjectsEncounteredWithoutPermissions`, since "since startup" is implied as this can't be persisted.

Comment (rekmarks) [reply packages/controllers/src/subject-metadata/SubjectMetadataController.ts]: Done in: d4cdb68

**Review by Gudahtt — APPROVED** [review]
LGTM!
